packages/editor/src/systems/ClickPlacementSystem.tsx

```diff
@@ -0,0 +1,280 @@
+import { MeshComponent } from '@etherealengine/spatial/src/renderer/components/MeshComponent'
+import React from 'react'
+
+export const ClickPlacementState = defineState({
+  name: 'ClickPlacementState',
+  initial: {
+    placementEntity: UndefinedEntity as Entity,
+    placedEntity: UndefinedEntity as Entity,
+    selectedAsset: null as string | null,
+    yawOffset: 0,
+    pitchOffset: 0,
+    rollOffset: 0,
+    materialCache: [] as [Mesh, Material][]
+  }
+})
+
+const PlacementModelReactor = (props: { placementEntity: Entity }) => {
+  const clickState = useState(getMutableState(ClickPlacementState))
+  const sceneState = useState(getMutableState(SceneState))
+  const placementModel = useOptionalComponent(props.placementEntity, ModelComponent)
+
+  useEffect(() => {
+    if (!placementModel) return
+    const sceneID = getModelSceneID(props.placementEntity)
+    if (!sceneState.scenes[sceneID]) return
+    iterateEntityNode(props.placementEntity, (entity) => {
+      const mesh = getOptionalComponent(entity, MeshComponent)
+      if (!mesh) return
+      const material = mesh.material as Material
+      clickState.materialCache.set((prev) => [...prev, [mesh, material]])
+      mesh.material = new HolographicMaterial({})
+    })
+  }, [placementModel?.scene, sceneState.scenes.keys])
+
+  return null
+}
+
+const objectLayerQuery = defineQuery([ObjectLayerComponents[ObjectLayers.Scene]])
+
+export const ClickPlacementSystem = defineSystem({
+  uuid: 'ee.studio.ClickPlacementSystem',
+  insert: { before: TransformSystem },
+  reactor: () => {
+    const clickState = useState(getMutableState(ClickPlacementState))
+    const editorState = useState(getMutableState(EditorHelperState))
+    const sceneState = useState(getMutableState(SceneState))
+
+    const renderers = defineQuery([RendererComponent])
+
+    const getParentEntity = () => {
+      if (!sceneState.scenes.value) return null
+      const sceneID = sceneState.scenes.keys[0] as SceneID
+      const scene = SceneState.getScene(sceneID)
+      if (!scene) return null
+      const entity = UUIDComponent.getEntityByUUID(scene.scene.root)
+      return entity
+    }
+
+    const createPlacementEntity = (parentEntity: Entity) => {
+      const placementEntity = createSceneEntity('Placement', parentEntity)
+      //removeComponent(placementEntity, SourceComponent)
+      return placementEntity
+    }
+
+    const clickListener = () => {
+      if (!clickState.selectedAsset.value) return
+      const parentEntity = getParentEntity()
+      if (!parentEntity) return
+      const placementEntity = clickState.placementEntity.value
+      if (!placementEntity) return
+      const sceneID = getComponent(parentEntity, SourceComponent)
+      setComponent(placementEntity, SourceComponent, sceneID)
+      setComponent(placementEntity, EntityTreeComponent, { parentEntity })
+      const snapshot = SceneSnapshotState.cloneCurrentSnapshot(sceneID)
```

Comment:
can we move this snapshot manipulation logic to EditorControlFunctions to keep it in the same context as the other ones?

Reply:
Yes this specific logic doesn't yet exist in the current editor control functions, since we are both changing the scene of the entity and adding it to the authoring layer at the same time